cpp/inc/bond/protocol/detail/protobuf_utils.h

```diff
@@ -0,0 +1,299 @@
+
+    template <typename Unused>
+    const std::string AttributeName<Unused>::pack = "ProtoPack";
+
+    inline Encoding ReadEncoding(const std::string& name, const Metadata& metadata)
+    {
+        auto it = metadata.attributes.find(name);
+        if (it == metadata.attributes.end())
+        {
+            return Unavailable<Encoding>();
+        }
+
+        if (std::strcmp(it->second.c_str(), "Fixed") == 0)
+        {
+            return Encoding::Fixed;
+        }
+
+        if (std::strcmp(it->second.c_str(), "ZigZag") == 0)
+        {
+            return Encoding::ZigZag;
+        }
+
+        InvalidEnumValueException(name.c_str(), "Encoding");
+    }
+
+    inline Encoding ReadEncoding(BondDataType type, const Metadata& metadata, const std::string& name)
+    {
+        switch (type)
+        {
+        case BT_FLOAT:
+        case BT_DOUBLE:
+            return Encoding::Fixed;
+
+        case BT_UINT8:
+        case BT_UINT16:
+        case BT_UINT32:
+        case BT_UINT64:
+        case BT_INT8:
+        case BT_INT16:
+        case BT_INT32:
+        case BT_INT64:
+            return ReadEncoding(name, metadata);
+
+        case BT_WSTRING:
+        case BT_SET:
+        case BT_MAP:
+            NotSupportedException("Wide string or nested set/map");
+
+        case BT_BOOL:
+        case BT_LIST:
+        case BT_STRING:
+        case BT_STRUCT:
+        default:
+            return Unavailable<Encoding>();
+        }
+    }
+
+    inline Encoding ReadEncoding(BondDataType type, const Metadata& metadata)
+    {
+        return ReadEncoding(type, metadata, AttributeName<>::encode);
+    }
+
+    inline Encoding ReadKeyEncoding(BondDataType type, const Metadata& metadata)
+    {
+        if (type == BT_FLOAT || type == BT_DOUBLE)
+        {
+            NotSupportedException("Float/double map key");
+        }
+
+        return ReadEncoding(type, metadata, AttributeName<>::encode_key);
+    }
+
+    inline Encoding ReadValueEncoding(BondDataType type, const Metadata& metadata)
+    {
+        return ReadEncoding(type, metadata, AttributeName<>::encode_value);
+    }
+
+    inline Packing ReadPacking(const std::string& name, const Metadata& metadata)
+    {
+        auto it = metadata.attributes.find(name);
+        if (it == metadata.attributes.end())
+        {
+            return Unavailable<Packing>();
+        }
+
+        if (std::strcmp(it->second.c_str(), "False") == 0)
+        {
+            return Packing::False;
+        }
+
+        InvalidEnumValueException(name.c_str(), "Packing");
+    }
+
+    inline Packing ReadPacking(BondDataType type, const Metadata& metadata)
+    {
+        switch (type)
+        {
+        case BT_STRING:
+        case BT_STRUCT:
+        case BT_LIST:
+            return Packing::False;
+
+        case BT_WSTRING:
+        case BT_SET:
+        case BT_MAP:
+            NotSupportedException("Wide string or nested set/map");
```

Comment:
Why are nested sets/maps a problem?

Reply:
Cannot be expressed in protobuf (e.g. `vector<vector<bool>>` cannot be mapped to `repeated repeated bool`), unless we box it.